File: apps/mobile/android/app/src/main/java/tech/zseven/rish/tasks/TaskExperience.kt
package tech.zseven.rish.tasks

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import tech.zseven.rish.MainActivity
import org.json.JSONArray
import org.json.JSONObject
import java.util.Locale

/** Main-thread-only display/lifecycle coordinator. A saved display never authorizes work. */
internal object TaskExperience {
    const val ONGOING_ID = 9101
    const val RETIRED_START_ID = 9102
    const val RUNNING_CHANNEL = "rish.task.running.v1"
    private const val ALERT_CHANNEL = "rish.task.alerts.v1"
    const val OPEN = "tech.zseven.rish.OPEN_TASK"
    const val CANCEL = "tech.zseven.rish.CANCEL_TASK"
    // Process-lifetime Application only; never retain an Activity here.
    private lateinit var context: android.app.Application
    private val prefs get() = context.getSharedPreferences("rish.task.experience.v1", Context.MODE_PRIVATE)
    private val notifications get() = context.getSystemService(NotificationManager::class.java)
    var emit: ((String) -> Unit)? = null
    var foreground = false
        private set
    private var visible: String? = null
    private var locale = Locale.getDefault().toLanguageTag()
    private var current: Run? = null
    private val queued = linkedMapOf<String, JSONObject>()
    private val notified = mutableSetOf<String>()
    private var preferences = defaults()
    var backgroundAdmission = "notRequested"
        private set
    data class Run(val id: String, val conversation: String, var phase: String, val startedAt: Long = System.currentTimeMillis(), var stopRequested: Boolean = false, var interrupted: Boolean = false)
    fun activeRun(): Run? = current

    fun initialize(app: Context) {
        if (::context.isInitialized) return
        context = app.applicationContext as android.app.Application
        preferences = try { validatePreferences(JSONObject(prefs.getString("preferences", "{}")!!)) } catch (_: Exception) { defaults() }
        // Cancel only our orphan progress card. Completion alerts remain readable.
        notifications.cancel(ONGOING_ID)
        notifications.cancel(RETIRED_START_ID)
        prefs.edit().remove("activeRun").apply()
        if (Build.VERSION.SDK_INT >= 26) {
            notifications.createNotificationChannel(NotificationChannel(RUNNING_CHANNEL, "Rish task status", NotificationManager.IMPORTANCE_LOW).apply {
                description = "Status of a user-started Rish task"; setShowBadge(false)
            })
            notifications.createNotificationChannel(NotificationChannel(ALERT_CHANNEL, "Rish task alerts", NotificationManager.IMPORTANCE_DEFAULT).apply {
                description = "Opt-in completion, failure and attention alerts"
            })
        }
    }
    private fun defaults() = JSONObject().put("completed", false).put("failed", false).put("attention", false)
        .put("liveActivity", true).put("background", false).put("muted", JSONArray())
    private fun validId(id: String): Boolean = id.isNotEmpty() && id.toByteArray(Charsets.UTF_8).size <= 256 && !id.contains('\u0000')
    private fun validatePreferences(raw: JSONObject): JSONObject {
        val keys = setOf("completed", "failed", "attention", "liveActivity", "background", "muted")
        require(raw.keys().asSequence().toSet() == keys)
        for (key in keys - "muted") require(raw.get(key) is Boolean)
        val muted = raw.getJSONArray("muted")
        require(muted.length() <= 500)
        for (index in 0 until muted.length()) require(muted.get(index) is String && validId(muted.getString(index)))
        return JSONObject(raw.toString())
    }
    private fun enabled(key: String) = preferences.getBoolean(key)
    private fun muted(id: String): Boolean = preferences.getJSONArray("muted").let { list -> (0 until list.length()).any { list.getString(it) == id } }
    fun permissionRequested() { prefs.edit().putBoolean("permissionRequested", true).apply() }
    fun authorization(): String {
        if (notifications.areNotificationsEnabled() && (Build.VERSION.SDK_INT < 33 || context.checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED)) return "authorized"
        return if (Build.VERSION.SDK_INT >= 33 && !prefs.getBoolean("permissionRequested", false)) "notDetermined" else "denied"
    }
    fun settings(): JSONObject = JSONObject().put("available", true).put("notifications", authorization())
        .put("liveActivitiesAvailable", false).put("backgroundAvailable", true)
        .put("backgroundAdmission", backgroundAdmission).put("preferences", JSONObject(preferences.toString()))

    fun handle(request: JSONObject): Any {
        require(TaskPolicy.validSchema(request.opt("schema_version")))
        return when (request.getString("op")) {
            "settings" -> settings()
            "preferences" -> {
                val next = validatePreferences(request.getJSONObject("preferences"))
                check(prefs.edit().putString("preferences", next.toString()).commit())
                preferences = next
                if (!enabled("background")) {
                    context.stopService(Intent(context, TaskForegroundService::class.java))
                    if (!foreground) current?.let { requestCancel(it.id, it.conversation, true) }
                }
                refreshProgress(); settings()
            }
            "visible" -> {
                visible = request.opt("conversationId") as? String
                (request.opt("locale") as? String)?.let { if (it in setOf("zh-CN", "en-US")) locale = it }
                true
            }
            "drain" -> JSONArray(queued.values.toList()).also { queued.clear() }
            "begin" -> {
                val runId = request.getString("runId"); val conversation = request.getString("conversationId")
                require(validId(runId) && validId(conversation))
                if (current?.id != runId) {
                    current?.let { finish(it.id, "cancelled") }
                    current = Run(runId, conversation, "preparing")
                    prefs.edit().putString("activeRun", runId).apply()
                    notified.clear(); backgroundAdmission = "notRequested"
                    if (enabled("background") && foreground) {
                        val intent = Intent(context, TaskForegroundService::class.java).putExtra("runId", runId)
                        try {
                            backgroundAdmission = "submitted"
                            if (Build.VERSION.SDK_INT >= 26) context.startForegroundService(intent) else context.startService(intent)
                        } catch (_: RuntimeException) { backgroundAdmission = "unavailable" }
                    }
                    refreshProgress()
                    if (!foreground && backgroundAdmission != "accepted") requestCancel(runId, conversation, true)
                }
                true
            }
            "update" -> {
                val phase = request.getString("phase")
                require(phase in setOf("preparing", "persistence_pending", "starting", "sending", "approval_pending", "executing", "recovering", "cancelling", "finalizing", "retryable", "resume_available", "commit_pending", "blocked"))
                val run = current
                if (run != null && TaskPolicy.owns(request.optString("runId"), run.id)) {
                    if (phase != "approval_pending") notified.remove("attention")
                    run.phase = phase; refreshProgress()
                    if (phase == "approval_pending") alert("attention")
                    true
                } else false
            }
            "end" -> finish(request.getString("runId"), request.getString("status"))
            else -> throw IllegalArgumentException("Unknown task operation")
        }
    }
    fun resume() { foreground = true }
    fun pause() {
        foreground = false; visible = null
        current?.let { if (backgroundAdmission !in setOf("accepted", "submitted")) requestCancel(it.id, it.conversation, true) }
    }
    fun backgroundAccepted(runId: String) {
        if (TaskPolicy.owns(runId, current?.id)) backgroundAdmission = "accepted"
    }
    fun backgroundLost(runId: String) {
        val run = current ?: return
        if (!TaskPolicy.owns(runId, run.id)) return
        backgroundAdmission = "unavailable"
        if (!foreground) requestCancel(run.id, run.conversation, true)
    }
    fun requestCancel(runId: String, conversation: String, interrupted: Boolean = false): Boolean {
        val run = current ?: return false
        if (!TaskPolicy.owns(runId, run.id) || run.conversation != conversation || run.stopRequested) return false
        run.stopRequested = true; run.interrupted = interrupted; run.phase = "cancelling"
        queue("cancel", runId, conversation)
        refreshProgress()
        return true
    }
    fun open(intent: Intent?) {
        if (intent?.action != OPEN) return
        val run = intent.getStringExtra("runId") ?: return
        val conversation = intent.getStringExtra("conversationId") ?: return
        if (validId(run) && validId(conversation)) queue("open", run, conversation)
    }
    private fun queue(action: String, runId: String, conversation: String) {
        val event = JSONObject().put("schema_version", 1).put("action", action).put("runId", runId).put("conversationId", conversation)
        queued["$action:$runId"] = event
        while (queued.size > 32) queued.remove(queued.keys.first())
        try { emit?.invoke(event.toString()) } catch (_: RuntimeException) {
            // Keep the action queued if React is being torn down.
        }
    }
    private fun finish(runId: String, status: String): Boolean {
        val run = current ?: return false
        if (!TaskPolicy.owns(runId, run.id)) return false
        if (status == "completed") alert("completed") else if (status != "cancelled" || run.interrupted) alert("failed")
        current = null
        prefs.edit().remove("activeRun").apply()
        notifications.cancel(ONGOING_ID)
        context.stopService(Intent(context, TaskForegroundService::class.java))
        return true
    }
    private fun label(en: String, zh: String) = if (locale.startsWith("zh")) zh else en
    private fun stage(phase: String): String = when (phase) {
        "preparing", "starting" -> label("Preparing", "正在准备")
        "sending" -> label("Waiting for model", "等待模型回复")
        "executing" -> label("Running tools", "正在执行工具")
        "approval_pending" -> label("Needs your attention", "需要你处理")
        "finalizing" -> label("Saving result", "正在保存结果")
        "cancelling" -> label("Stopping safely", "正在安全停止")
        else -> label("Return to Rish to continue", "返回 Rish 继续处理")
    }
    private fun openIntent(run: Run): PendingIntent {
        val intent = Intent(context, MainActivity::class.java).setAction(OPEN)
            .setData(Uri.Builder().scheme("rish-task").authority("open").appendPath(run.id).build())
            .putExtra("runId", run.id).putExtra("conversationId", run.conversation)
            .addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP)
        return PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
    }
    fun retiredStartNotification(): Notification = builder(RUNNING_CHANNEL, label("Finishing task", "结束任务")).setOnlyAlertOnce(true).build()
    fun progressNotification(): Notification? {
        val run = current ?: return null
        val cancel = Intent(context, TaskActionReceiver::class.java).setAction(CANCEL)
            .setData(Uri.Builder().scheme("rish-task").authority("cancel").appendPath(run.id).build())
            .putExtra("runId", run.id).putExtra("conversationId", run.conversation)
        val pending = PendingIntent.getBroadcast(context, 0, cancel, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        return builder(RUNNING_CHANNEL, stage(run.phase)).setOngoing(true).setOnlyAlertOnce(true)
            .setWhen(run.startedAt).setShowWhen(true).setUsesChronometer(true)
            .setContentIntent(openIntent(run)).addAction(android.R.drawable.ic_menu_close_clear_cancel, label("Stop", "停止"), pending).build()
    }
    private fun builder(channel: String, body: String): Notification.Builder {
        val builder = if (Build.VERSION.SDK_INT >= 26) Notification.Builder(context, channel) else Notification.Builder(context)
        return builder.setSmallIcon(android.R.drawable.stat_notify_sync).setContentTitle("Rish").setContentText(body)
            .setVisibility(Notification.VISIBILITY_PRIVATE).setCategory(Notification.CATEGORY_PROGRESS)
            .setShowWhen(false)
    }
    private fun refreshProgress() {
        if (current == null) return
        if ((enabled("liveActivity") || backgroundAdmission in setOf("accepted", "submitted")) && authorization() == "authorized") {
            try { notifications.notify(ONGOING_ID, progressNotification()) } catch (_: SecurityException) { }
        } else if (backgroundAdmission !in setOf("accepted", "submitted")) notifications.cancel(ONGOING_ID)
    }
    private fun alert(kind: String) {
        val run = current ?: return
        if (!TaskPolicy.allowsAlert(kind, enabled(kind), muted(run.conversation), foreground && visible == run.conversation,
                kind in notified, true) || authorization() != "authorized") return
        val body = when (kind) {
            "completed" -> label("Task complete", "任务已完成")
            "attention" -> label("Needs your attention", "需要你处理")
            else -> label("Task needs review", "任务需要检查")
        }
        val notification = builder(ALERT_CHANNEL, body).setCategory(Notification.CATEGORY_STATUS).setAutoCancel(true)
            .setContentIntent(openIntent(run)).build()
        try { notifications.notify(run.id, listOf("completed", "failed", "attention").indexOf(kind) + 1, notification); notified.add(kind) }
        catch (_: SecurityException) { }
    }
}

File: apps/mobile/ios/Rish/TaskExperience.swift
import UIKit
import UserNotifications
import BackgroundTasks
import ActivityKit
import React

struct TaskPreferences: Codable {
  var completed = false
  var failed = false
  var attention = false
  var liveActivity = true
  var background = false
  var muted: [String] = []
}

@MainActor
final class TaskExperience: NSObject, UNUserNotificationCenterDelegate {
  static let shared = TaskExperience()
  static var prefix: String { (Bundle.main.bundleIdentifier ?? "tech.zseven.rish") + ".task" }
  static let preferenceKey = "rish.task.preferences.v1"
  private var preferences = TaskPreferences()
  private var current: [String: String]?
  private var startedAt = Date()
  private var visibleConversation: String?
  private var language: String?
  private var notified: Set<String> = []
  private var activity: Any?
  private var backgroundTask: Any?
  private var backgroundIdentifier: String?
  private var completedBackgroundIdentifiers: [String] = []
  private var registrationCount = 0
  private var backgroundAdmission = "notRequested"
  private var supportsBackground: Bool {
    if #available(iOS 26.0, *) { return registrationCount < 256 }
    return false
  }
  private var lease: UIBackgroundTaskIdentifier = .invalid
  private var actions: [[String: String]] = []
  var emit: (([String: String]) -> Void)?

  func configure() {
    if let data = UserDefaults.standard.data(forKey: Self.preferenceKey),
       let saved = try? JSONDecoder().decode(TaskPreferences.self, from: data) { preferences = saved }
    UNUserNotificationCenter.current().delegate = self
    // A display checkpoint never authorizes execution after launch.
    if let old = UserDefaults.standard.string(forKey: "rish.task.background.id") {
      BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: old)
      UserDefaults.standard.removeObject(forKey: "rish.task.background.id")
    }
    if #available(iOS 16.2, *) {
      Task { for old in Activity<RishTaskAttributes>.activities { await old.end(nil, dismissalPolicy: .immediate) } }
    }
    NotificationCenter.default.addObserver(forName: UIApplication.didEnterBackgroundNotification, object: nil, queue: .main) { _ in
      Task { @MainActor in self.enterBackground() }
    }
    NotificationCenter.default.addObserver(forName: UIApplication.didBecomeActiveNotification, object: nil, queue: .main) { _ in
      Task { @MainActor in
        if self.lease != .invalid { UIApplication.shared.endBackgroundTask(self.lease); self.lease = .invalid }
      }
    }
  }

  private var chinese: Bool { (language ?? Locale.preferredLanguages.first ?? "en").hasPrefix("zh") }
  private func text(_ en: String, _ zh: String) -> String { chinese ? zh : en }
  private func stage(_ phase: String) -> String {
    switch phase {
    case "preparing", "starting": return text("Preparing", "正在准备")
    case "sending": return text("Waiting for model", "等待模型回复")
    case "approval_pending": return text("Needs your attention", "需要你处理")
    case "executing": return text("Running tools", "正在执行工具")
    case "recovering": return text("Checking saved state", "核对已保存状态")
    case "cancelling": return text("Stopping safely", "正在安全停止")
    case "finalizing": return text("Saving result", "正在保存结果")
    case "completed": return text("Task complete", "任务已完成")
    case "cancelled": return text("Task stopped", "任务已停止")
    default: return text("Return to Rish to continue", "返回 Rish 继续处理")
    }
  }

  func handle(_ input: String) async throws -> Any {
    guard input.utf8.count < 32768,
          let data = input.data(using: .utf8),
          let request = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          request["schema_version"] as? Int == 1, let op = request["op"] as? String else { throw invalid() }
    switch op {
    case "inspection":
      guard Bundle.main.bundleIdentifier?.hasPrefix("dev.zseven.rish.taskacceptance") == true else { throw invalid() }
      var activityActive = false
      if #available(iOS 16.2, *), let live = activity as? Activity<RishTaskAttributes> { activityActive = live.activityState == .active }
      return ["visibleConversation": visibleConversation ?? "", "running": current != nil,
        "phase": current?["phase"] ?? "idle", "interrupted": current?["interrupted"] == "true", "stopRequested": current?["stopRequested"] == "true",
        "liveActivityActive": activityActive, "backgroundAccepted": backgroundTask != nil, "backgroundAdmission": backgroundAdmission,
        "applicationActive": UIApplication.shared.applicationState == .active]
    case "settings": return await settings()
    case "preferences":
      guard let raw = request["preferences"] as? [String: Any] else { throw invalid() }
      let value = try JSONDecoder().decode(TaskPreferences.self, from: JSONSerialization.data(withJSONObject: raw))
      guard value.muted.count <= 500, value.muted.allSatisfy(validID) else { throw invalid() }
      preferences = value
      UserDefaults.standard.set(try JSONEncoder().encode(value), forKey: Self.preferenceKey)
      if !value.liveActivity { await endActivity() }
      if !value.background {
        finishBackground(success: false)
        if UIApplication.shared.applicationState != .active { enterBackground() }
      }
      return await settings()
    case "permission":
      _ = try await UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound])
      return await settings()
    case "openSettings":
      if let url = URL(string: UIApplication.openSettingsURLString) { await UIApplication.shared.open(url) }
      return true
    case "visible":
      visibleConversation = request["conversationId"] as? String
      if let locale = request["locale"] as? String, ["zh-CN", "en-US"].contains(locale) { language = locale }
      return true
    case "drain": let result = actions; actions.removeAll(); return result
    case "begin":
      guard let run = request["runId"] as? String, validID(run),
            let conversation = request["conversationId"] as? String, validID(conversation) else { throw invalid() }
      if current?["runId"] == run { return true }
      await endActivity(); finishBackground(success: false)
      current = ["runId": run, "conversationId": conversation, "phase": "preparing"]
      startedAt = Date(); notified.removeAll(); backgroundAdmission = "notRequested"
      // Request only while a user-initiated operation owns the foreground.
      if preferences.background, UIApplication.shared.applicationState == .active, #available(iOS 26.0, *), supportsBackground {
        let identifier = Self.prefix + "." + UUID().uuidString
        let task = BGContinuedProcessingTaskRequest(identifier: identifier, title: "Rish", subtitle: stage("preparing"))
        task.strategy = .fail
        // Wildcards belong in Info.plist; registration requires the concrete
        // identifier. Handlers do not retain a run or resume work. Bound the
        // number of process-lifetime registrations to avoid accumulating them.
        let registered = BGTaskScheduler.shared.register(forTaskWithIdentifier: identifier, using: .main) { task in
          Task { @MainActor in self.accept(task) }
        }
        registrationCount += 1
        if registered {
          do {
            backgroundIdentifier = identifier
            try BGTaskScheduler.shared.submit(task)
            backgroundAdmission = "submitted"
            UserDefaults.standard.set(identifier, forKey: "rish.task.background.id")
          } catch {
            backgroundIdentifier = nil
            backgroundAdmission = "unavailable"
          }
        } else { backgroundAdmission = "registrationRejected" }
      }
      if preferences.liveActivity, #available(iOS 16.2, *), ActivityAuthorizationInfo().areActivitiesEnabled {
        activity = try? Activity.request(attributes: RishTaskAttributes(conversationId: conversation, runId: run, startedAt: startedAt, locale: chinese ? "zh-CN" : "en-US"),
          content: ActivityContent(state: .init(stage: stage("preparing"), finished: false), staleDate: Date().addingTimeInterval(90)), pushType: nil)
      }
      if UIApplication.shared.applicationState != .active { enterBackground() }
      return true
    case "update":
      guard let phase = request["phase"] as? String,
            ["preparing", "persistence_pending", "starting", "sending", "approval_pending", "executing", "recovering", "cancelling", "finalizing", "retryable", "resume_available", "commit_pending", "blocked"].contains(phase) else { throw invalid() }
      guard current?["runId"] == request["runId"] as? String, current != nil else { return false }
      if phase != "approval_pending" { notified.remove("attention") }
      current?["phase"] = phase
      if #available(iOS 26.0, *), let task = backgroundTask as? BGContinuedProcessingTask {
        // Count actual stage transitions; never manufacture timed progress.
        task.progress.completedUnitCount += 1
        task.updateTitle("Rish", subtitle: stage(phase))
      }
      if #available(iOS 16.2, *), let live = activity as? Activity<RishTaskAttributes> {
        await live.update(ActivityContent(state: .init(stage: stage(phase), finished: false), staleDate: Date().addingTimeInterval(90)))
      }
      if phase == "approval_pending" { await notify("attention") }
      return true
    case "end":
      guard current?["runId"] == request["runId"] as? String, current != nil else { return false }
      let status = request["status"] as? String ?? "blocked"
      if status == "completed" { await notify("completed") }
      else if status != "cancelled" || current?["interrupted"] == "true" { await notify("failed") }
      await endActivity(); finishBackground(success: status == "completed")
      current = nil
      return true
    default: throw invalid()
    }
  }

  private func validID(_ value: String) -> Bool { !value.isEmpty && value.utf8.count <= 256 && !value.contains("\0") }
  private func invalid() -> NSError { NSError(domain: "RishTaskExperience", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid task request"]) }
  private func settings() async -> [String: Any] {
    let status = await UNUserNotificationCenter.current().notificationSettings()
    let authorization: String
    switch status.authorizationStatus {
    case .authorized, .provisional, .ephemeral: authorization = "authorized"
    case .denied: authorization = "denied"
    default: authorization = "notDetermined"
    }
    var live = false
    if #available(iOS 16.2, *) { live = ActivityAuthorizationInfo().areActivitiesEnabled }
    return ["available": true, "notifications": authorization, "liveActivitiesAvailable": live,
            "backgroundAvailable": supportsBackground, "backgroundAdmission": backgroundAdmission,
            "preferences": (try? JSONSerialization.jsonObject(with: JSONEncoder().encode(preferences))) ?? [:]]
  }

  private func notify(_ kind: String) async {
    guard let current else { return }
    let settings = await UNUserNotificationCenter.current().notificationSettings()
    let enabled = kind == "completed" ? preferences.completed : kind == "failed" ? preferences.failed : preferences.attention
    guard TaskAlertPolicy.allows(kind: kind, enabled: enabled,
      muted: preferences.muted.contains(current["conversationId"]!),
      foreground: UIApplication.shared.applicationState == .active,
      viewingConversation: visibleConversation == current["conversationId"],
      delivered: notified.contains(kind), sameRun: self.current?["runId"] == current["runId"]),
      [.authorized, .provisional, .ephemeral].contains(settings.authorizationStatus) else { return }
    let content = UNMutableNotificationContent()
    content.title = "Rish"
    content.body = kind == "completed" ? stage("completed") : kind == "attention" ? stage("approval_pending") : text("Task needs review", "任务需要检查")
    content.sound = .default
    content.threadIdentifier = current["conversationId"]!
    content.userInfo = ["conversationId": current["conversationId"]!, "runId": current["runId"]!]
    do {
      try await UNUserNotificationCenter.current().add(UNNotificationRequest(identifier: current["runId"]! + "." + kind, content: content, trigger: nil))
      notified.insert(kind)
    } catch { /* Display delivery must not change durable task outcome. */ }
  }

  private func enterBackground() {
    guard let run = current?["runId"], backgroundTask == nil else { return }
    lease = UIApplication.shared.beginBackgroundTask(withName: "Rish task checkpoint") {
      Task { @MainActor in
        guard self.current?["runId"] == run else { return }
        self.requestCancel(runId: run); self.finishBackground(success: false)
      }
    }
    if lease == .invalid { requestCancel(runId: run) }
  }
  @available(iOS 26.0, *) private func accept(_ task: BGTask) {
    guard let task = task as? BGContinuedProcessingTask, task.identifier == backgroundIdentifier, current != nil else {
      task.setTaskCompleted(success: completedBackgroundIdentifiers.contains(task.identifier)); return
    }
    backgroundTask = task
    backgroundAdmission = "accepted"
    // An agent has no known total number of steps. Keep its progress
    // indeterminate instead of inventing a completion percentage.
    task.progress.totalUnitCount = -1
    task.progress.completedUnitCount = 0
    let ownerRun = current!["runId"]!
    let ownerIdentifier = task.identifier
    task.expirationHandler = { Task { @MainActor in
      guard self.current?["runId"] == ownerRun, self.backgroundIdentifier == ownerIdentifier else { return }
      self.requestCancel(runId: ownerRun)
      // Keep a short cleanup lease until the existing controller confirms its
      // durable cancellation. Never acknowledge completion before that flush.
      if self.lease == .invalid {
        self.lease = UIApplication.shared.beginBackgroundTask(withName: "Rish cancellation checkpoint") {
          Task { @MainActor in
            guard self.current?["runId"] == ownerRun else { return }
            self.finishBackground(success: false)
          }
        }
      }
      if self.lease == .invalid { self.finishBackground(success: false) }
    } }
    // The system continued-processing presentation already owns this task.
    Task {
      guard self.current?["runId"] == ownerRun else { return }
      await endActivity()
    }
  }
  private func finishBackground(success: Bool) {
    if #available(iOS 26.0, *), let task = backgroundTask as? BGContinuedProcessingTask {
      if success {
        let completed = max(1, task.progress.completedUnitCount)
        task.progress.totalUnitCount = completed; task.progress.completedUnitCount = completed
      }
      task.setTaskCompleted(success: success)
    }
    backgroundTask = nil
    if let id = backgroundIdentifier {
      if success { completedBackgroundIdentifiers = Array((completedBackgroundIdentifiers + [id]).suffix(32)) }
      BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: id)
    }
    backgroundIdentifier = nil
    UserDefaults.standard.removeObject(forKey: "rish.task.background.id")
    if lease != .invalid { UIApplication.shared.endBackgroundTask(lease); lease = .invalid }
  }
  private func endActivity() async {
    let previous = activity
    activity = nil
    if #available(iOS 16.2, *), let live = previous as? Activity<RishTaskAttributes> { await live.end(nil, dismissalPolicy: .immediate) }
  }
  func cancelFromActivity(runId: String, conversationId: String) {
    guard current?["conversationId"] == conversationId else { return }
    requestCancel(runId: runId, interrupted: false)
  }
  private func requestCancel(runId: String, interrupted: Bool = true) {
    guard current?["runId"] == runId else { return }
    current?["interrupted"] = interrupted ? "true" : "false"
    current?["stopRequested"] = "true"
    guard let current else { return }
    queue(["action": "cancel", "conversationId": current["conversationId"]!, "runId": current["runId"]!])
    Task {
      guard self.current?["runId"] == runId else { return }
      await endActivity()
    }
  }
  private func queue(_ action: [String: String]) {
    if !actions.contains(action) { actions.append(action); actions = Array(actions.suffix(32)) }
    emit?(action)
  }
  func open(_ url: URL) -> Bool {
    guard url.scheme == "rish", url.host == "task", let parts = URLComponents(url: url, resolvingAgainstBaseURL: false),
          let conversation = parts.queryItems?.first(where: { $0.name == "conversation" })?.value, validID(conversation),
          let run = parts.queryItems?.first(where: { $0.name == "run" })?.value, validID(run) else { return false }
    queue(["action": "open", "conversationId": conversation, "runId": run]); return true
  }
  nonisolated func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification) async -> UNNotificationPresentationOptions {
    await MainActor.run {
      let conversation = notification.request.content.userInfo["conversationId"] as? String
      return conversation == self.visibleConversation ? [] : [.banner, .sound]
    }
  }
  nonisolated func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse) async {
    guard let conversation = response.notification.request.content.userInfo["conversationId"] as? String,
          let run = response.notification.request.content.userInfo["runId"] as? String else { return }
    await MainActor.run { self.queue(["action": "open", "conversationId": conversation, "runId": run]) }
  }
}

@objc(RishTaskExperience)
class RishTaskExperienceModule: RCTEventEmitter {
  override static func requiresMainQueueSetup() -> Bool { true }
  override func supportedEvents() -> [String]! { ["RishTaskAction"] }
  override func startObserving() {
    Task { @MainActor in
      TaskExperience.shared.emit = { [weak self] action in
        var envelope: [String: Any] = action; envelope["schema_version"] = 1
        if let data = try? JSONSerialization.data(withJSONObject: envelope), let json = String(data: data, encoding: .utf8) {
          self?.sendEvent(withName: "RishTaskAction", body: json)
        }
      }
    }
  }
  override func stopObserving() { Task { @MainActor in TaskExperience.shared.emit = nil } }
  @objc(handle:resolver:rejecter:)
  func handle(_ input: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    Task { @MainActor in
      do {
        let value = try await TaskExperience.shared.handle(input)
        let data = try JSONSerialization.data(withJSONObject: ["schema_version": 1, "ok": true, "value": value])
        resolve(String(data: data, encoding: .utf8))
      } catch { reject("E_TASK_EXPERIENCE", "Task service unavailable", nil) }
    }
  }
}
